docs/resource/README.md

```diff
@@ -102,7 +102,105 @@ as specified in the [Resource SDK specification](https://github.com/open-telemet
+implementations SHOULD use Version 5 and MUST use the following UUID as the namespace:
+`4d63009a-8d0f-11ee-aad7-4c796ed8e320`.
+
+UUIDs are typically recommended, as we only need an opaque yet reproducible value for
+the purposes of identifying a service instance. Similar to what can be seen in the man page for the
+[`/etc/machine-id`](https://www.freedesktop.org/software/systemd/man/machine-id.html) file,
+the underlying data, such as pod name and namespace should be treated as
+confidential by this algorithm, being the user's choice to expose it or not via another resource attribute.
+
+When a UUID v5 is generated, the input MUST be prefixed with
+`${telemetry.sdk.name}.${telemetry.sdk.language}.${service.namespace}.${service.name}`, followed by the
+the workload identifier, which should tentatively stable. This means that the same service yields the
+same UUID if the same identifier (`host.id`, `/etc/machine-id`, and so on) remains the same. It would still
+yield different results for different services on the same host or namespace. When no namespaces or equivalent fields
+are available, the prefix MUST then be `${telemetry.sdk.name}.${telemetry.sdk.language}.${service.name}`.
+
+Users running their services on platforms such as Kubernetes are encouraged to explicitly set the
+`service.instance.id` using their existing automation, or by setting a value that can be used for a
+consistent value, such as `container.id`. Similarly, users of application servers such as `unicorn`
+are encouraged to set the `service.instance.id` on a per-worker basis.
+
+SDKs MUST use the following algorithm when generating `service.instance.id`:
+
+- If the user has provided a `service.instance.id`, via environment
+  variable, configuration or custom resource detection, this will
+  always be used and honored over generated IDs.
+- When any of the below combinations of resource attribute are provided, they MUST be used as the input
+  for generating a UUID v5 following the prefix mentioned above. The values within each combination MUST be separated with dots:
+  * `container.id`, resulting in the input `${telemetry.sdk.name}.${telemetry.sdk.language}.${service.namespace}.${service.name}.${container.id}`
+  * `k8s.namespace.name`/`k8s.pod.name`/`k8s.container.name`, resulting in the input
+  `${telemetry.sdk.name}.${telemetry.sdk.language}.${k8s.namespace.name}.${service.name}.${k8s.pod.name}.${k8s.container.name}`
+  * `host.id`, resulting in the input `${telemetry.sdk.name}.${telemetry.sdk.language}.${service.namespace}.${service.name}.${host.id}`
+- When the SDK is running in an environment where a `/etc/machine-id`
+  (see [MACHINE-ID(5)](https://www.freedesktop.org/software/systemd/man/machine-id.html))
+  is available, the machine-id should be used in the input for generating a UUID v5:
+  `${telemetry.sdk.name}.${telemetry.sdk.language}.${service.namespace}.${service.name}.${machine.id}`
```

Comment:
Would be great if all the algorithms could be based on resource attributes. That way, the SDK code generating `service.instance.id` could just run as the last step before the resource is resolved, inspect all the attributes that have been provided so far, and select the correct algorithm. Not fond of the idea of having the implementation have to go and inspect files in the OS.

Reply:
I guess if machine id is important enough to include as a component of the prefix which generates `service.instance.id`, why isn't it important enough to define as a resource attribute?

Reply:
machine-id, along with other candidates, can be used as a source for host.id. Specifically for machine-id, it can't be used as-is in a resource attribute, as its man-page defines that it should be treated as confidential information.